File: Web-App/test/kontofilter-mitgeben.test.js
/**
 * Was der Kontofilter neu lädt, muss den Kontofilter auch mitschicken.
 *
 * ── Der Befund ──────────────────────────────────────────────────────────────
 * `onScopeChange('gallery')` in 02-gallery.js ruft zwei Dinge auf:
 *
 *     if (view === 'gallery')  { loadGallery(); loadStats(); }
 *
 * `loadGallery()` schickte den Filter mit, `loadStats()` nicht — es rief
 * schlicht `api('GET', '/v1/stats')`. Die Route wertet `accounts` aus
 * (routes/api_v1/misc.ts, parseScopeMode), bekam ihn aber nie.
 *
 * Wer auf „nur meine" stellte, sah darunter eine gefilterte Liste und darüber
 * weiter die Zahlen des ganzen Haushalts — Sets, Teile, Minifiguren,
 * Anleitungen. Am Telefon stimmten beide: Die App schickt accounts an
 * /v1/stats seit jeher.
 *
 * ── Warum das ein BEWEIS ist und keine Vermutung ────────────────────────────
 * Der Aufruf von `loadStats()` steht IM Filterwechsel. Etwas neu zu laden, das
 * sich durch den Wechsel gar nicht ändern kann, ergibt nur einen Sinn, wenn es
 * sich ändern sollte. Die Absicht steht also im Code selbst — sie war nur
 * nicht umgesetzt.
 *
 * ── Der zweite Fund, und warum der Test dafür umgebaut wurde ────────────────
 * Die erste Fassung fragte je LADEFUNKTION: „steht irgendwo in ihrer
 * Aufrufkette ein scopeQuery()?" Damit war sie blind für
 * `enrichGalleryWithPrices()`: Es hängt am Ende von `loadGallery()`, das den
 * Filter über galleryParams() sehr wohl mitgibt — der Aufruf auf
 * /v1/finance/pnl darin aber nicht. Die Kette trug ihn, der Aufruf nicht.
 *
 * Geprüft wird deshalb JE AUFRUF. Das ist die Regel, die beide Fälle trifft.
 *
 * ── Warum gesucht und nicht aufgezählt ──────────────────────────────────────
 * Drei Seiten werden gelesen: WAS der Filterwechsel neu lädt (aus
 * onScopeChange), WELCHE v1-Routen `accounts` überhaupt auswerten (aus den
 * Router-Dateien) und OB jeder Aufruf dorthin den Filter mitgibt. Ein neuer
 * Endpunkt, eine neue Ansicht, ein neuer Zwischenaufruf — alles ist von selbst
 * mitgeprüft. Eine Liste hier wäre eine zweite Wahrheit, die beim nächsten
 * Umbau still veraltet.
 */
const test = require('node:test');
const assert = require('node:assert/strict');
const fs = require('node:fs');
const path = require('node:path');
const { ohneKommentare } = require('./helpers/sources');

const JS = path.join(__dirname, '..', 'public', 'js');

/** Alle Frontend-Module als ein Text — die Ladefunktionen liegen verstreut. */
function alleModule() {
  const teile = {};
  for (const f of fs.readdirSync(JS)) {
    if (!f.endsWith('.js') || f === 'app.bundle.js') continue;
    teile[f] = ohneKommentare(fs.readFileSync(path.join(JS, f), 'utf8'));
  }
  return teile;
}

/**
 * Rumpf einer Funktion ab ihrem Namen bis zur schliessenden Klammer.
 * Geklammert gezählt statt über ein festes Zeilenfenster: Ein gewachsener
 * Erklärkommentar hat in diesem Baum schon mehrfach genau so eine Prüfung
 * gebrochen (siehe helpers/sources.js).
 */
function rumpf(src, kopf) {
  const i = src.indexOf(kopf);
  if (i < 0) return null;
  // Erst die Parameterliste überspringen. Die erste `{` nach dem Namen kann
  // ein VORGABEWERT sein — `loadGallery(opts = {})` hat genau die Form, und
  // ein Zähler, der dort beginnt, ist nach zwei Zeichen wieder bei null und
  // liefert einen 43 Zeichen langen „Rumpf". Der Test wäre dann rot gewesen,
  // weil er nichts gefunden hat, nicht weil etwas fehlt.
  let k = src.indexOf('(', i);
  if (k < 0) return null;
  let klammern = 1;
  for (k++; k < src.length && klammern > 0; k++) {
    if (src[k] === '(') klammern++;
    else if (src[k] === ')') klammern--;
  }
  let j = src.indexOf('{', k);
  if (j < 0) return null;
  let tiefe = 1;
  for (j++; j < src.length && tiefe > 0; j++) {
    if (src[j] === '{') tiefe++;
    else if (src[j] === '}') tiefe--;
  }
  return src.slice(i, j);
}

test('jeder Abruf hinter dem Kontofilter gibt den Kontofilter mit', () => {
  const module = alleModule();
  const gallery = module['02-gallery.js'];
  assert.ok(gallery, '02-gallery.js nicht gefunden');

  const wechsel = rumpf(gallery, 'export function onScopeChange(');
  assert.ok(wechsel, 'onScopeChange() nicht gefunden — umbenannt?');

  // Was der Wechsel neu lädt: jeder Aufruf `loadXyz()` in seinem Rumpf.
  const lader = [...new Set([...wechsel.matchAll(/\b(load[A-Z]\w*)\s*\(/g)].map(m => m[1]))];
  // Selbstbeweis: Findet das Muster keine Lader, wäre unten nichts zu prüfen
  // und der Test grün, ohne etwas geprüft zu haben. Vier Ansichten (Galerie,
  // Teile, Minifiguren, Finanzen) laden mindestens je einen.
  assert.ok(lader.length >= 4,
    `Nur ${lader.length} Lader im Filterwechsel gefunden — Muster veraltet?`);

  /** Rumpf einer Funktion, in welchem Modul auch immer sie steht. */
  function suche(name) {
    for (const [datei, src] of Object.entries(module)) {
      const k = rumpf(src, `export async function ${name}(`)
             ?? rumpf(src, `export function ${name}(`)
             ?? rumpf(src, `async function ${name}(`)
             ?? rumpf(src, `function ${name}(`);
      if (k) return { datei, koerper: k };
    }
    return null;
  }

  /**
   * Welche v1-Routen werten `accounts` aus? Aus den Router-Dateien gelesen,
   * nicht aufgezählt — eine Liste hier wäre eine zweite Wahrheit, die beim
   * nächsten neuen Endpunkt still veraltet.
   */
  function kontenbewusst() {
    const wurzel = path.join(__dirname, '..', 'routes', 'api_v1');
    const menge = new Set();
    for (const f of fs.readdirSync(wurzel)) {
      if (!f.endsWith('.ts') || f === 'index.ts' || f === 'middleware.ts') continue;
      const src = ohneKommentare(fs.readFileSync(path.join(wurzel, f), 'utf8'));
      for (const m of src.matchAll(/router\.(get|post|put|delete)\(\s*'([^']+)'([\s\S]*?)\n\}\);/g))
        if (m[3].includes('query.accounts')) menge.add(m[2].split('/:')[0].replace(/\/+$/, ''));
    }
    return menge;
  }

  /**
   * Trägt dieser Aufruf den Filter?
   *
   * Geprüft wird JE AUFRUF, nicht je Ladefunktion. Die erste Fassung dieses
   * Tests fragte „steht irgendwo in der Aufrufkette ein scopeQuery()?" — und
   * war damit blind für genau den zweiten Fund: `loadGallery()` baut den
   * Filter über galleryParams(), ruft am Ende aber
   * `enrichGalleryWithPrices()`, das /v1/finance/pnl OHNE Filter holte. Die
   * Kette trug ihn, der Aufruf nicht.
   */
  function aufrufeOhneFilter(name, konten, tiefe = 0, gesehen = new Set()) {
    if (tiefe > 3 || gesehen.has(name)) return [];
    gesehen.add(name);
    const treffer = suche(name);
    if (!treffer) return [];
    const fehlt = [];
    for (const m of treffer.koerper.matchAll(/api\(\s*'GET'\s*,\s*[`'"]([^`'"]*?)[`'"\s)]/g)) {
      const pfad = m[1].split('?')[0].split('${')[0].replace(/\/+$/, '');
      if (!konten.has(pfad.replace(/^\/v1/, ''))) continue;
      // Der Filter darf im Aufruf selbst stehen (scopeQuery/scopeMode) oder in
      // der Params-Funktion, deren Ergebnis angehängt wird.
      const zeile = treffer.koerper.slice(m.index, m.index + 220);
      if (/scopeQuery\(|scopeMode\(|accounts/.test(zeile)) continue;
      const viaParams = [...zeile.matchAll(/\b(\w*[Pp]arams)\s*\(/g)]
        .some(p => { const t2 = suche(p[1]); return t2 && /addScopeParam\(|scopeQuery\(|scopeMode\(/.test(t2.koerper); });
      if (viaParams) continue;
      fehlt.push(`${treffer.datei}: ${name}() -> ${pfad}`);
    }
    for (const m of treffer.koerper.matchAll(/\b([a-z]\w*)\s*\(/g))
      if (m[1] !== name) fehlt.push(...aufrufeOhneFilter(m[1], konten, tiefe + 1, gesehen));
    return fehlt;
  }

  const konten = kontenbewusst();
  // 22 Routen werten accounts aus; auf den Pfadstamm vor dem ersten
  // Platzhalter zusammengefasst bleiben 14. Die Grenze soll „gar nichts
  // gefunden" fangen, nicht das Entfernen eines Endpunkts verbieten.
  assert.ok(konten.size >= 12,
    `Nur ${konten.size} kontenbewusste Routen gefunden — Muster veraltet?`);

  const ohneFilter = [];
  let geprueft = 0;
  for (const name of lader) {
    if (!suche(name)) continue;
    geprueft++;
    ohneFilter.push(...aufrufeOhneFilter(name, konten));
  }

  const offen = [...new Set(ohneFilter)].sort();
  assert.deepEqual(offen, [],
    'Diese Aufrufe hängen am Kontofilter, schicken ihn aber nicht mit:\n  ' +
    offen.join('\n  ') +
    '\nSie werden beim Wechsel neu geladen und liefern trotzdem jedesmal ' +
    'dasselbe. Der Nutzer sieht dann eine gefilterte Liste neben ungefilterten ' +
    'Zahlen — genau das war bei loadStats() und bei enrichGalleryWithPrices() der Fall.');
});

File: Android-App/app/src/test/java/ch/brickinventoryapp/KontofilterMitgebenTest.kt
package ch.brickinventoryapp

import org.junit.Test

/**
 * Was der Kontofilter neu laedt, muss das Blickfeld auch mitschicken.
 *
 * ── Woher die Regel kommt ───────────────────────────────────────────────────
 * In der Webapp ist genau das ZWEIMAL schiefgegangen. `onScopeChange('gallery')`
 * laedt dort `loadGallery()` und `loadStats()` neu — und `loadStats()` rief
 * schlicht `api('GET', '/v1/stats')`. Wer auf „nur meine" stellte, sah eine
 * gefilterte Liste ueber ungefilterten Zahlen. Beim zweiten Fall
 * (enrichGalleryWithPrices) trug die AUFRUFKETTE den Filter, der Aufruf darin
 * aber nicht.
 *
 * ── Was hier geprueft wird ──────────────────────────────────────────────────
 * Die App macht es an allen 13 Stellen richtig. Diese Pruefung haelt das fest:
 * Jeder `repo.…`-Aufruf, den ein Lader hinter [setScope] erreicht, muss das
 * Blickfeld mitgeben — entweder direkt als `scopeFor(…)` oder ueber eine
 * Variable, die im selben Rumpf daraus belegt wurde (so macht es
 * loadValuation() mit `acc`).
 *
 * ── Warum gesucht und nicht aufgezaehlt ─────────────────────────────────────
 * WELCHE Lader der Filterwechsel anstoesst, steht in setScope() selbst und
 * wird von dort gelesen. Kommt eine Ansicht dazu oder laedt ein Wechsel etwas
 * Weiteres nach, ist es von selbst mitgeprueft.
 */
class KontofilterMitgebenTest {

    @Test
    fun `jeder Abruf hinter dem Kontofilter fuehrt das Blickfeld mit`() {
        val quellen = Quellen.unter("ui").associate { it.name to Quellen.ohneKommentare(it.readText()) }

        val wechsel = rumpf("setScope", quellen)
        assert(wechsel != null) { "setScope() nicht gefunden — umbenannt?" }

        // Was der Wechsel neu laedt.
        val lader = Regex("""\b(load[A-Z]\w*)\s*\(""").findAll(wechsel!!.second)
            .map { it.groupValues[1] }.toSet()
        // Selbstbeweis: Findet das Muster keine Lader, waere unten nichts zu
        // pruefen und der Test gruen, ohne etwas geprueft zu haben.
        assert(lader.size >= 6) { "Nur ${lader.size} Lader in setScope() gefunden — Muster veraltet?" }

        val fehlend = mutableListOf<String>()
        var geprueft = 0

        fun pruefe(name: String, tiefe: Int, gesehen: MutableSet<String>) {
            if (tiefe > 3 || !gesehen.add(name)) return
            val (datei, koerper) = rumpf(name, quellen) ?: return
            // Variablen, die im selben Rumpf aus scopeFor(...) belegt werden —
            // loadValuation() holt das Blickfeld einmal und reicht es dreimal
            // weiter. Ein Fenster ab der Aufrufstelle sieht das nicht, weil die
            // Zuweisung DAVOR steht.
            val blickfeldVars = Regex("""val (\w+)\s*=\s*scopeFor\(""")
                .findAll(koerper).map { it.groupValues[1] }.toSet()

            for (t in Regex("""repo\.(\w+)\.(\w+)\s*\(""").findAll(koerper)) {
                geprueft++
                val ab = t.range.first
                val bis = minOf(koerper.length, ab + 260)
                val stelle = koerper.substring(ab, bis)
                val traegt = stelle.contains("scopeFor(") || stelle.contains("accounts") ||
                    blickfeldVars.any { v -> Regex("""\W$v\W""").containsMatchIn(stelle) }
                if (!traegt) {
                    val zeile = koerper.substring(0, ab).count { it == '\n' } + 1
                    fehlend += "$datei: $name() -> repo.${t.groupValues[1]}.${t.groupValues[2]}() (Zeile ~$zeile im Rumpf)"
                }
            }
            for (m in Regex("""\b(load[A-Z]\w*)\s*\(""").findAll(koerper))
                if (m.groupValues[1] != name) pruefe(m.groupValues[1], tiefe + 1, gesehen)
        }

        for (l in lader) pruefe(l, 0, mutableSetOf())

        // Zweiter Selbstbeweis: Untersucht die Suche gar keine Aufrufe, waere
        // „nichts gefunden" wertlos. Genau dieser Fall ist beim Nachmessen
        // aufgetreten — ein falscher Pfadvergleich, null untersuchte Aufrufe,
        // und trotzdem eine gruene Meldung.
        assert(geprueft >= 10) {
            "Nur $geprueft repo-Aufrufe untersucht — dann sagt ein leeres Ergebnis nichts."
        }

        assert(fehlend.isEmpty()) {
            "Diese Abrufe haengen am Kontofilter, geben ihn aber nicht mit:\n  " +
                fehlend.joinToString("\n  ") +
                "\nSie werden beim Wechsel neu geladen und liefern trotzdem jedesmal " +
                "dasselbe — eine gefilterte Liste neben ungefilterten Zahlen."
        }
    }

    private companion object {
        /** Rumpf einer Funktion, in welcher Datei auch immer sie steht. */
        fun rumpf(name: String, quellen: Map<String, String>): Pair<String, String>? {
            for ((datei, src) in quellen) {
                val m = Regex("""fun (?:MainViewModel\.)?${Regex.escape(name)}\s*\(""").find(src) ?: continue
                // Erst die Parameterliste ueberspringen: Die erste '{' nach dem
                // Namen kann ein Vorgabewert sein.
                var i = src.indexOf('(', m.range.first)
                var tiefe = 1
                i++
                while (i < src.length && tiefe > 0) {
                    if (src[i] == '(') tiefe++ else if (src[i] == ')') tiefe--
                    i++
                }
                val start = src.indexOf('{', i)
                if (start < 0) continue
                var t2 = 1
                var e = start + 1
                while (e < src.length && t2 > 0) {
                    if (src[e] == '{') t2++ else if (src[e] == '}') t2--
                    e++
                }
                return datei to src.substring(m.range.first, e)
            }
            return null
        }
    }
}
